File: algorithms/160.c
#include <stdio.h>

struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB)
{
    if ( !headA || !headB )
        return NULL;
    struct ListNode *p1, *p2;
    struct ListNode *tailA = headA, *tailB = headB;
    while ( (*tailA).next != NULL )
        tailA = (*tailA).next;
    while ( (*tailB).next != NULL )
        tailB = (*tailB).next;
    if ( tailA != tailB )
        return NULL;
    (*tailA).next = headA;
    p1 = (*headB).next;
    p2 = (*p1).next;
    while ( p1 != p2 )
    {
        p1 = (*p1).next;
        p2 = (*p2).next;
        p2 = (*p2).next;
    }
    p2 = headB;
    while ( p1 != p2 )
    {
        p1 = (*p1).next;
        p2 = (*p2).next;
    }
    (*tailA).next = NULL;
    return p1;
}

int main()
{
    return 0;
}

File: algorithms/python/160.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if not headA or not headB: return None
        tail = headA
        while tail.next:
            tail = tail.next
        tail.next = headB

        slow, fast = headA, headA
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                break
        else:
            tail.next = None
            return None

        fast = headA
        while slow != fast:
            slow = slow.next
            fast = fast.next

        tail.next = None
        return fast
